klippy/webhooks.py

```diff
@@ -0,0 +1,245 @@
+    def __init__(self, webhooks, printer):
+        self.webhooks = webhooks
+        self.reactor = printer.get_reactor()
+
+        # Klippy Connection
+        self.fd_handler = self.mutex = None
+        self.is_server_connected = False
+        self.partial_data = ""
+        is_fileoutput = (printer.get_start_args().get('debugoutput')
+                         is not None)
+        if is_fileoutput:
+            # Do not try to connect in klippy batch mode
+            return
+        self.socket = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
+        self.socket.setblocking(0)
+        try:
+            self.socket.connect(SOCKET_LOCATION)
+        except socket.error:
+            logging.debug(
+                "ServerConnection: Moonraker server not detected")
+            return
+        logging.debug("ServerConnection: Moonraker connection established")
+        self.is_server_connected = True
+        self.fd_handler = self.reactor.register_fd(
+            self.socket.fileno(), self.process_received)
+        self.mutex = self.reactor.mutex()
+        printer.register_event_handler('klippy:disconnect', self.close_socket)
+
+    def close_socket(self):
+        if self.is_server_connected:
+            logging.info("ServerConnection: lost connection to Moonraker")
+            self.is_server_connected = False
+            self.reactor.unregister_fd(self.fd_handler)
+            try:
+                self.socket.close()
+            except socket.error:
+                pass
+
+    def is_connected(self):
+        return self.is_server_connected
+
+    def process_received(self, eventtime):
+        try:
+            data = self.socket.recv(4096)
+        except socket.error as e:
+            # If bad file descriptor allow connection to be
+            # closed by the data check
+            if e.errno == 9:
+                data = ''
+            else:
+                return
+        if data == '':
+            # Socket Closed
+            self.close_socket()
+            return
+        requests = data.split('\x00')
+        requests[0] = self.partial_data + requests[0]
+        self.partial_data = requests.pop()
+        for req in requests:
+            logging.debug(
+                "ServerConnection: Request received from Moonraker %s" % (req))
+            try:
+                decoded_req = json_loads_byteified(req)
+                self._process_request(decoded_req)
+            except Exception:
+                logging.exception(
+                    "ServerConnection: Error processing Server Request %s"
+                    % (req))
+
+    def _process_request(self, req):
+        web_request = WebRequest(req)
+        try:
+            func = self.webhooks.get_callback(
+                web_request.get_path())
+            func(web_request)
+        except WebRequestError as e:
+            web_request.set_error(e)
+        except Exception as e:
+            web_request.set_error(WebRequestError(e.message))
```

Comment:
If the code catches a generic error here, I think it should write the underlying error to the log and transition to a shutdown state (as is done in the similar handler in gcode.py).  It's annoying to get a shutdown on a code error, but the alternative could be much worse.